torch/csrc/jit/tensorexpr/mem_dependency_checker.h

```diff
@@ -0,0 +1,402 @@
+  std::vector<const Expr*> getIndices() const;
+
+  // Establishes a dependency or dependent relationship with another access.
+  void addDependency(const std::shared_ptr<AccessInfo>& write);
+  void addDependent(const std::shared_ptr<AccessInfo>& read);
+
+  // helper for checking dependencies.
+  bool hasDependency(const std::shared_ptr<AccessInfo>& info) const;
+
+  // Returns the set of all nodes that are direct (immediate) dependencies of
+  // this access.
+  DependencySet getDirectDependencies();
+  // likewise, returns all nodes that directly depend on this one.
+  DependencySet getDirectDependents();
+
+  // Returns the full list of all nodes in the graph that this access depends
+  // on, and all nodes they depend on, and so forth, back to the inputs.
+  DependencySet getIndirectDependencies();
+  // likewise, returns the full list of all nodes that depend on this node, and
+  // all nodes that depend on those nodes and so on down to the outputs.
+  DependencySet getIndirectDependents();
+
+  // Does this access represent a read of memory (Load, ReduceOp, Call, etc).
+  bool isRead() const;
+  // Does this access represent a write of memory (Store, etc).
+  bool isWrite() const;
+
+  // Helpers for dumping accesses in various formats.
+  void print() const;
+  void dumpDOT(std::ostream& os) const;
+  const char* AccessTypeColour() const;
+
+ private:
+  size_t id_;
+  AccessType type_;
+  const Stmt* stmt_;
+  const Expr* expr_;
+  const Var* var_;
+  IndexBounds bounds_;
+
+  // Yes these should be sorted.
+  std::map<size_t, std::shared_ptr<AccessInfo>> dependencies_;
+  std::map<size_t, std::shared_ptr<AccessInfo>> dependents_;
+};
+
+using VarBoundMap = std::unordered_map<const Var*, Bound>;
+
+/* MemDepedencyChecker analyses a IR fragment and builds a dependency graph of
+ * accesses contained within.
+ *
+ * It's possible to retrieve the entire graph in node-object form, or can be
+ * used as an oracle for answering dependency questions. e.g:
+ *
+ *  analyzer.hasIndirectDependency(BufA, BufB); or,
+ *  analyzer.hasDirectDependency(LoadA, StoreB);
+ */
+class TORCH_API MemDependencyChecker : public IRVisitor {
+  struct Scope;
+
+ public:
+  MemDependencyChecker();
+  MemDependencyChecker(
+      const std::unordered_set<const Buf*>& inputs,
+      const std::unordered_set<const Buf*>& outputs);
+  MemDependencyChecker(
+      const std::vector<BufHandle>& inputs,
+      const std::vector<BufHandle>& outputs);
+
+  virtual ~MemDependencyChecker() {}
+
+  // Whether or not to allow loop execution order to influence dependency
+  // calculation. If the loop may later be parallelized you don't want this.
+  bool allowLoopExecutionOrderAnalysis(bool allow = true);
+
+  // Dependency Checking API.
+  // The goal is to have enough overloads here so you don't really have to think
+  // about it.
+
+  // Returns true if any read in A has a direct dependence on a write in B.
```

Comment:
Why only check for ReadAfterWrite dependency (read in A depending on write in B)? We could also have a WriteAfterWrite (output) dependency, or a WriteAfterRead dependency (which we might have to check to ensure that some transformation like loop fusion is valid). How do you suggest we check for those?

Reply:
This is a good point, what do you suggest?

Can you give an example of where WriteAfterWrite matters?

Reply:
I can't think of a use-case that could come up from the lowering of ops we have right now. May be if we reuse buffers?

But, in general, loops could have such dependencies and transformations like loop fusion will have to make sure they are not violated.

Here is a contrived example:
```
for i in 100
    a[i] =  x[i];
for i in 50
    a[i + 1] = y[i];
```
It is invalid to fuse these two loops in this case, because we have WriteAfterWrite dependency.

Reply:
You already have the tools to check for other dependencies here. It is only a matter of figuring out a suitable API. Here are couple of suggestions:
1. Have the checking methods return an `enum` denoting what kind of dependency is present.
2. Have different APIs for each kind of dependency.

Also, you can probably handle it as a separate PR; this PR is already doing a lot.

Reply:
Yeah I'm thinking that also, I'll land this and we can iterate on this API.